### ISSUE TITLE
Determine if column exists in expression context

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#8019]
### Problem description

I've seen similar requests to this, but I often have expression-generating functions that are dependent on the presence of other columns and, should those columns not be present, I need to create them via an expression. Something like the following:

```python
import polars as pl

df1 = pl.DataFrame({
    'a': [1, 2, 3]
})
df2 = pl.DataFrame({
    'b': [1, 2, 3]
})

expr = pl.if_col('a').then(pl.col('a')+1).otherwise(0).alias('b')

df1.select(expr)  # 'b' is [2, 3, 4]
df2.select(expr) # 'b' is [0, 0, 0
```

**Comment by ghuls** [comment]
You can use custom function with `pipe` to get this behavior:

```python
In [100]: def if_col(df: pl.DataFrame, column_name, expr_found, expr_not_found) -> pl.DataFrame:
     ...:     if column_name in df.schema:
     ...:         return df.with_columns(expr_found)
     ...:     else:
     ...:         return df.with_columns(expr_not_found)
     ...: 

In [101]: df1.pipe(if_col, "a", (pl.col('a') + 1).alias("b"), pl.lit(0).alias("b"))
Out[101]: 
shape: (3, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 1   ┆ 2   │
│ 2   ┆ 3   │
│ 3   ┆ 4   │
└─────┴─────┘

In [102]: df2.pipe(if_col, "a", (pl.col('a') + 1).alias("b"), pl.lit(0).alias("b"))
Out[102]: 
shape: (3, 1)
┌─────┐
│ b   │
│ --- │
│ i32 │
╞═════╡
│ 0   │
│ 0   │
│ 0   │
└─────┘
```

**Comment by mcrumiller** [comment]
@ghuls thanks--but I think this still isn't good enough. The purpose here is, if starting in an expression context for which you might not have access to the original dataframe, can you generate an expression conditional on the column being present on the dataframe?

**Comment by ghuls** [comment]
I don't think it can be done in the expression context as there the schema is not known yet (expressions are not bound to a DataFrame/LazyFrame instance), so you can't check for the presence or absence of of columns.

With custom functions and `pipe`, you have access to the schema of the DataFrame/LazyFrame instance.

**Comment by StefanBRas** [comment]
I have the same issue, where I implement custom expressions in a `ExprNameSpace` where the schema isn't available.

Some cases can be solved by using that `pl.col("^EXACT_COLUMN_NAME$")` will return an empty selection if the column doesn't exist, and then using this in `polars.coalesce`.

This obviously doesn't make for very readable or intuitive code so I too would like this feature.

**Comment by mcrumiller** [comment]
@StefanBRas that's a great solution that I've never thought about:

```python
df.with_columns(
    pl.coalesce(
        pl.col("^EXACT_COLUMN_NAME$").apply(found_logic),
        pl.col(other_col_names).apply(not_found_logic)
)
```

**Comment by ritchie46** [comment]
> I don't think it can be done in the expression context as there the schema is not known yet (expressions are not bound to a DataFrame/LazyFrame instance), so you can't check for the presence or absence of of columns.
> 
> With custom functions and `pipe`, you have access to the schema of the DataFrame/LazyFrame instance.

This. It is not possible. `pipe` gives you access to the `LazyFrame` in question and should be used for this.
